Make top-level code variables `@MainActor @preconcurrency`

etcwilde opened the pull request:
This patch gets the semantics roughly where we want them. The top-level code context is set to run on the main actor and the top-level variables are `@MainActor @preconcurrency`. This pairing reduces the number of hops between actors for top-level code, provides some level of safety, and keeps the code breakage fairly minimal.

The next PR contains `await` detection to turn the top-level code into an async context once the first use of `await` is used in the top-level. If no `await`s are used, the top-level should behave _exactly_ the same as it does today to keep current scripts working.

etcwilde committed:
Top-level variables are on the MainActor

This is an interesting patch that may raise some eyebrows. The machinery
we have now can completely handle arbitrary global-actor isolation on
the global variables. I'm forcing the variables to be on the main actor
because in an ideal future, the variables won't be global and will
actually be local to the implicit `async_Main` function, where they will
be isolated to the main actor by virtue of the `async_Main` function
being run by the main actor executor. For that to work, we can't really
allow top-level variables to have global actor isolation since local
variables can't have global actor isolation. So while there is no
technical reason we can't handle it now, to avoid source-breaking
changes in the future, I'm removing the ability to explicitly declare a
global actor on top-level global variables and forcing them all to be on
the MainActor implicitly.

etcwilde committed:
Top-level code is executed on the main actor

This patch sets the actor isolation of the top-level code contexts to be
on the main actor. This holds since the implicit `async_Main` function
generated is executed on the main actor executor.

etcwilde committed:
Update toplevel global actor isolation test

Since we're forcing all of the variables at the top-level to be main
actor isolated, there were some changes to where actor-hops were are to
happen. Additionally, I had to pull of the actor isolation on `a` since
we don't allow explicit global actor isolation. Given this, it doesn't
really make sense to keep `b` around. As a nice change though, we can
call synchronous functions directly in top level code, passing variables
directly.

etcwilde committed:
Add @preconcurrency to toplevel vars

Top-level global variables should have the `@predatesConcurrency
@MainActor` behaviour. This allows them to be somewhat concurrency safe
while still working with old code.

DougGregor commented on lib/AST/Decl.cpp:
```diff
@@ -733,10 +733,15 @@ bool Decl::predatesConcurrency() const {
+  // Variables declared in top-level code are @_predatesConcurrency
+  if (const VarDecl *var = dyn_cast<VarDecl>(this)) {
+    return getASTContext().LangOpts.EnableExperimentalAsyncTopLevel &&
+           var->isTopLevelGlobal();
```
Should we also check for Swift version < 6 here?

etcwilde replied:
I think that should get picked up by `contextRequiresStrictConcurrencyChecking`, which is used by `adjustVarTypeForConcurrecy` to strip off the actor from the type.
I'll throw it in just to ensure that nothing leaks past though.

etcwilde commented:
@swift-ci please test

etcwilde committed:
Test predatesConcurrency is only applied in Swift 5

Ensure that the predates concurrency behavior is only applied for Swift
5 and not for Swift 6.

etcwilde commented:
@swift-ci please test

etcwilde commented:
@swift-ci please test

ktoso commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -4741,7 +4741,9 @@ ERROR(global_actor_non_unsafe_init,none,
       "non-final class %0 cannot be a global actor", (DeclName))
-      
+ERROR(global_actor_top_level_var,none,
+      "top-level code variables cannot have a global actor", ())
```
add a test to cover this? there doesn't seem to be one

etcwilde replied:
Done 

https://github.com/apple/swift/pull/40998/files#diff-0b3cf9b84143abf54688a2441cb7d2425cb6819e229078be58f610ba2f09c1c7R6-R8

ktoso reviewed (approved):
Looks good to me, perhaps add one test to cover the error this also added

etcwilde committed:
Test disallowing global actors on top-level vars

Adding a test to ensure that we emit an error message if there are
global actors on top-level variables.

etcwilde commented:
@swift-ci please test

etcwilde commented:
It looks like all of the API's have changed... D:
Missing `predatesConcurrency` and `swift::ActorIsolation::forGlobalActor`!? Interesting.

stevapple commented:
`predatesConcurrency` has been changed into `preconcurrency` in #40680

etcwilde commented:
@swift-ci please test
